hack4impact-calpoly/five-cities-food-pantry
# Client Page Implementation, Updated Client Schema

## Developer: Aidan Nesbitt

Closes #43 #54 

### Pull Request Summary

1. Added functionality to the check in button so that it updates the client state in the database (and locally).
2. Updated the client schema to have an "isCheckedOff" attribute.
3. Updated each document in the database to have this new attribute ^.
4. Created an endpoint to (PUT) update the client document.
5. Created a banner that displays if the user is currently checked in.
6. Update the dummy data on the clientPage to display a specific client's data based on the URL parameter.

### Modifications

- Created an API endpoint to handle the PUT request to update the "isCheckedOff" state at: "api/clients/check-in/[userId]"
		- Added a function to handle PUT requests in the route.ts
- Modified clientDashboard.tsx, clientDashboard.css
		- This is where the most changes were created. The check in button now changes appearance when clicked, calls to the backend are made when appropriate, added HTML to display checked In state, added the dialog HTML and CSS. Made a GET request to populate specific client data in the clientPage instead of dummy data.
- clientSchema.ts: updated the schema and interface to include the "isCheckedOff" attribute.
- updated each document in the database to have the "isCheckedOff" attribute
- Modified: "api/clients/[userId]" to declare household member and authorized member models so that the client documents can be populated with their respective househodl/authorized members.

### Testing Considerations

- Client Information does render slowly, I believe this may be due to the needing to populate the householdMem and authMem lists within the client document... Is there a better way to handle this?
- Test that navigating to a specific client, by putting their userId in the URL, works properly.
- Test that checking in the user (with the button) correctly updates the 'isCheckedOff' attribute of that client document in the database.
- Check that if you update a clients checked in state, you can navigate away from the specific client page, then return and the state is correctly displayed.
- Ensure that client information is displayed correctly in all of the different fields: name, age, phone number, etc. 
		- ensure the age function calculates correctly
- I did update the calculate age function within the clientschema validation to match the age function created for the clientDashboard page: ensure this new age function doesn't break anything? (Probably fine haha).
- Try to test every user interaction: check in then click button again but select don't check out, check in then do check out...

### Other Notes
- I did not create an error message if the check in fails, but I added comments to indicate where these messages may be added (if deemed needed).
- Have we created the front end functionality to add householdMem/authMem _ids to their respective lists within their headofHousehold client documents, when creating a client profile/document in the database? For example, David Greens household members are Mark Smith and Alice Johnson. Is there front end functionality to put their unique _ids within David Greens householdMem list? Also, do we have functionality to create their own documents within the respective householdMem or authMem collections in the db?

### Pull Request Checklist


### Screenshots/Screencast

<img width="1380" alt="Screenshot 2024-05-18 at 2 29 37 PM" src="https://github.com/hack4impact-calpoly/five-cities-food-pantry/assets/75648657/eb7b93e7-de0c-42c9-b46a-a6cbb9e290a7">
<img width="1371" alt="Screenshot 2024-05-18 at 2 29 30 PM" src="https://github.com/hack4impact-calpoly/five-cities-food-pantry/assets/75648657/706c5ffc-ef68-4413-93fb-76c265d50ec8">
